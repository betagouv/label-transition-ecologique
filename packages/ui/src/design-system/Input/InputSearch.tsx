import classNames from 'classnames';
import {useRef} from 'react';
import {useDebouncedCallback} from 'use-debounce';
import {InputBase, InputBaseProps} from './InputBase';

export type InputSearchProps = Omit<InputBaseProps, 'icon' | 'type'> & {
  /** Fait apparaître un picto "chargement" à la place du picto "recherche" */
  isLoading?: boolean;
  /** Valeur en ms du délai entre deux appels à `onSearch` lors de la saisie. */
  debounce?: number;
  /** Appelée pour déclencher la recherche. Reçoit la valeur saisie dans le champ. */
  onSearch: (value: string) => void;
};

/**
 * Affiche un champ de recherche.
 * Un bouton permet de déclencher la recherche. Le bouton est désactivé si le champ est vide.
 * La recherche se déclenche également automatiquement (avec un `debounce`) lors de la saisie
 */
export const InputSearch = ({
  className,
  isLoading = false,
  debounce = 500,
  onChange,
  onSearch,
  ...remainingProps
}: InputSearchProps) => {
  const inputRef = useRef<HTMLInputElement>(null);
  const disabled = !(remainingProps.value || inputRef.current?.value);

  /** Debounce les appels à `onSearch` */
  const handleDebouncedInputChange = useDebouncedCallback(v => {
    onSearch(v);
  }, debounce);

  return (
    <InputBase
      className={classNames(
        'search-reset:appearance-none search-reset:w-5 search-reset:h-5 search-reset:bg-close-circle-fill',
        className
      )}
      type="search"
      ref={inputRef}
      onChange={e => {
        onChange?.(e);
        handleDebouncedInputChange(e.target.value);
      }}
      icon={{
        buttonProps: {
          disabled,
          icon: isLoading ? 'loader-4-line animate-spin' : 'search-line',
          onClick: () => handleDebouncedInputChange(inputRef.current.value),
          title: 'Rechercher',
        },
      }}
      {...remainingProps}
    />
  );
};
