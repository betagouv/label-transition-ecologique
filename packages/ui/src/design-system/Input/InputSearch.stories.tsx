import {Meta, StoryObj} from '@storybook/react';
import {action} from '@storybook/addon-actions';
import {InputSearch} from './InputSearch';
import {useState} from 'react';

const meta: Meta<typeof InputSearch> = {
  title: 'Design System/InputSearch',
  component: InputSearch,
  render: args => {
    const [value, setValue] = useState(args.value || '');
    return (
      <InputSearch
        {...args}
        value={value}
        onChange={e => setValue(e.target.value)}
        onSearch={action('onSearch')}
      />
    );
  },
};

export default meta;

type Story = StoryObj<typeof InputSearch>;

/** Sans aucune props renseignée. */
export const Default: Story = {
  args: {},
};

/** Avec une valeur renseignée. */
export const AvecValeur: Story = {
  args: {
    value: 'valeur recherchée',
  },
};
