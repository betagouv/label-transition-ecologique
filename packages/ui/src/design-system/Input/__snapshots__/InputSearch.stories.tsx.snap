// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Design System/InputSearch AvecValeur smoke-test 1`] = `
<div class="inline-flex items-stretch border border-solid rounded-lg bg-grey-1 overflow-hidden focus-within:border-primary-5 border-grey-4">
  <input type="search"
         class="text-grey-8 px-4 outline-none text-md py-3 border-r border-solid border-grey-4"
         value="valeur recherchÃ©e"
  >
  <button title="Rechercher"
          class="rounded-lg w-fit border border-solid font-bold group cursor-pointer text-primary-7 hover:!text-primary-8 bg-white hover:!bg-primary-1 border-white hover:!border-primary-1 p-3.5 rounded-none border-none"
  >
    <div class="flex items-center gap-2">
      <span class="ri-search-line font-normal text-center h-5 w-5 text-[1.25rem] leading-[1.25rem] fill-primary-7 group-hover:fill-primary-8">
      </span>
    </div>
  </button>
</div>
`;

exports[`Design System/InputSearch Default smoke-test 1`] = `
<div class="inline-flex items-stretch border border-solid rounded-lg bg-grey-1 overflow-hidden focus-within:border-primary-5 border-grey-4">
  <input type="search"
         class="text-grey-8 px-4 outline-none text-md py-3 border-r border-solid border-grey-4"
         value
  >
  <button title="Rechercher"
          disabled
          class="rounded-lg w-fit border border-solid font-bold group cursor-not-allowed !text-primary-4 !bg-white !border-white p-3.5 rounded-none border-none"
  >
    <div class="flex items-center gap-2">
      <span class="ri-search-line font-normal text-center h-5 w-5 text-[1.25rem] leading-[1.25rem] fill-primary-4">
      </span>
    </div>
  </button>
</div>
`;
